src/Avalonia.Base/Input/TextInput/ITextInputMethodClient.cs

```diff
@@ -10,57 +8,81 @@ public interface ITextInputMethodClient
         /// The cursor rectangle relative to the TextViewVisual
         /// </summary>
         Rect CursorRectangle { get; }
+
         /// <summary>
         /// Should be fired when cursor rectangle is changed inside the TextViewVisual
         /// </summary>
         event EventHandler? CursorRectangleChanged;
+
         /// <summary>
         /// The visual that's showing the text
         /// </summary>
         Visual TextViewVisual { get; }
+
         /// <summary>
         /// Should be fired when text-hosting visual is changed
         /// </summary>
         event EventHandler? TextViewVisualChanged;
+
         /// <summary>
         /// Indicates if TextViewVisual is capable of displaying non-committed input on the cursor position
         /// </summary>
         bool SupportsPreedit { get; }
+
         /// <summary>
         /// Sets the non-committed input string
         /// </summary>
         void SetPreeditText(string? text);
 
-        /// <summary>
-        /// Sets the current composing region. This doesn't remove the composing text from the commited text.
-        /// </summary>
-        void SetComposingRegion(TextRange? region);
-
         /// <summary>
         /// Indicates if text input client is capable of providing the text around the cursor
         /// </summary>
         bool SupportsSurroundingText { get; }
+
         /// <summary>
         /// Returns the text around the cursor, usually the current paragraph, the cursor position inside that text and selection start position
         /// </summary>
-        TextInputMethodSurroundingText SurroundingText { get; }
+        string? SurroundingText { get; }
+
         /// <summary>
         /// Should be fired when surrounding text changed
         /// </summary>
         event EventHandler? SurroundingTextChanged;
 
         /// <summary>
-        /// Gets or sets a platform editable. Text and selection changes made in the editable are forwarded to the IM client.
+        /// Gets or sets he selection range within current surrounding text.
         /// </summary>
-        ITextEditable? TextEditable { get; set; }
+        TextSelection Selection { get; set; }
 
-        void SelectInSurroundingText(int start, int end);
-    }
+        /// <summary>
+        /// Should be fired when the selection changed
+        /// </summary>
+        event EventHandler? SelectionChanged;
 
-    public record struct TextInputMethodSurroundingText
-    {
-        public string Text { get; set; }
-        public int CursorOffset { get; set; }
-        public int AnchorOffset { get; set; }
+        /// <summary>
+        /// Inserts text at current cursor position.
+        /// </summary>
+        /// <param name="text">The inserted text</param>
+        /// <remarks>
+        /// Replaces currently selected text range.
+        /// </remarks>
+        void Insert(string? text);
```

Comment:
This skips the routed events infrastructure, so it prevents people from intercepting the event. This is a breaking change and doesn't match UWP/WPF.

Reply:
We can raise the event on TopLevel. We should still unify this implementation.

Reply:
Removed